Cleanup Renderer API to Remove Fixed-Function State

Web-eWorks opened the pull request:
Matrix state ~~must~~ did die!

Remove all OpenGL 2.x like matrix state methods from the renderer API, now the matrix state is the responsibility of the caller via MatrixTicket / StateTicket.

Still TODO is improving the viewport state tracking API, although that should be rather simple to fix via promoting the Viewport struct to the top-level Renderer API.

Needs a little dead-code-elimination because I wrote this between the hours of 3 and 4 AM, but should produce identical results to the previous system.

Ping @fluffyfreak - I recognize that `XXX` comment annotation and this was actually surprisingly simple to do, so there's gotta be a bug in it somewhere, right? *Right?*

Web-eWorks commented:
For the record, I'm not 100% happy with this solution as we're still keeping some state in the renderer, but it's still miles better than retaining unnecessary fixed-function semantics in our API. I'll revisit this when I get around to replacing individual uniforms with UBOs, as that will provide a decent opportunity to remove this kind of state "for free".

fluffyfreak commented on src/graphics/opengl/RendererGL.cpp:
```diff
@@ -603,11 +597,19 @@ namespace Graphics {
-		SetMatrixMode(MatrixMode::MODELVIEW);
-		LoadMatrix(m);
+
+		// SetMatrixMode(MatrixMode::MODELVIEW);
```
Remove instead of comment out

Web-eWorks replied:
This is that dead code elimination I was talking about - forgot to clean this up, but I still have one more commit to go, I'll clean it up then.

fluffyfreak commented on src/graphics/opengl/RendererGL.cpp:
```diff
@@ -696,7 +706,8 @@ namespace Graphics {
 	void RendererOGL::SetMaterialShaderTransforms(Material *m)
 	{
-		m->SetCommonUniforms(m_modelViewStack.top(), m_projectionStack.top());
+		//m->SetCommonUniforms(m_modelViewStack.top(), m_projectionStack.top());
```
Remove instead of comment out

fluffyfreak commented on src/graphics/Renderer.h:
```diff
@@ -145,59 +144,65 @@ namespace Graphics {
 		// is deleted, the renderer state is restored
 		// XXX state must die
-		class StateTicket {
+		class [[deprecated]] StateTicket
```
Deprecated?

Web-eWorks replied:
I deprecated the existing methods (to see the extent of their use) before removing them, this is an unintentional slip.

fluffyfreak commented:
It looks good and I agree with the direction 👍
When I wrote the Renderer state tracking stuff it was to avoid the OpenGL calls querying state, which are a performance killer especially on ATi/AMD drivers. 
It was never really with a view to refactoring it like this so this is good progress!

Web-eWorks commented:
While I'm at it, I'm adding explicit type conversion between the `float` and `double` versions of `matrix4x4` and completely removing the "helper function" that was previously involved. I'll also be removing the `double` overload of `Renderer::SetTransform` and fixing things up so they communicate better to the reader that there's a loss of precision involved in passing double-based matricies to the renderer.

Web-eWorks commented:
Fixed up Viewport handling, removed the `matrix4x4d` overload of `Renderer::SetTransform` (it makes it a little more messy when calling, but also explicitly clarifies that you're losing precision), and ran some clang-format over files that have slipped out of format compliance.

This also has the side-effect of majorly cleaning up `GasGiantJobs.cpp`, as we no longer need twelve lines to enter/leave orthographic state and we also no longer stomp renderer state!

Web-eWorks commented:
@fluffyfreak let me know if you want to review the new changes - if you're swamped currently I'm good to go ahead and merge, but I wanted to make sure I got your sign-off first!

fluffyfreak commented on src/core/GuiApplication.cpp:
```diff
@@ -25,7 +25,7 @@ void GuiApplication::BeginFrame()
 #endif
 	// TODO: render target size
-	m_renderer->SetViewport(0, 0, Graphics::GetScreenWidth(), Graphics::GetScreenHeight());
+	m_renderer->SetViewport({ 0, 0, Graphics::GetScreenWidth(), Graphics::GetScreenHeight() });
```
I need to test this kind of initialisation with Visual Studio builds, and non GCC builds basically.

Web-eWorks replied:
This *should* be standard-compliant list-initialization, but don't quote me on that. Appveyor didn't build because there's a merge conflict but I don't think it had problems with it.

Returning a list-initialized object (and this particular call) is a form of `copy-list-initialization`, which as far as I can tell is directly standard-compliant according to [the C++ reference](https://en.cppreference.com/w/cpp/language/list_initialization).

fluffyfreak replied:
Yeah this isn't your fault or a C++ standards issue 😉 it's a does-VS-conform-to-the-C++-standards issue, which with this particular list initialisation I have seen it fail before, so I want to double check it before it's merged

fluffyfreak commented:
I'm all good with this conceptually, I just need to check that it builds in Visual Studio.
I don't have time to do so tonight, but hopefully will tomorrow.

laarmen commented:
Yeah I don't think Andrew had any doubt about the code being standard
compliant, but more about VS knowing about the standard ;-)

That said, they've improved dramatically since the days of C++11.

On Wed, Sep 2, 2020 at 7:11 PM Webster Sheets <notifications@github.com>
wrote:

> *@Web-eWorks* commented on this pull request.
> ------------------------------
>
> In src/core/GuiApplication.cpp
> <https://github.com/pioneerspacesim/pioneer/pull/4947#discussion_r482232032>
> :
>
> > @@ -25,7 +25,7 @@ void GuiApplication::BeginFrame()
>  	m_renderer->SetRenderTarget(m_renderTarget);
>  #endif
>  	// TODO: render target size
> -	m_renderer->SetViewport(0, 0, Graphics::GetScreenWidth(), Graphics::GetScreenHeight());
> +	m_renderer->SetViewport({ 0, 0, Graphics::GetScreenWidth(), Graphics::GetScreenHeight() });
>
> This *should* be standard-compliant list-initialization, but don't quote
> me on that. Appveyor didn't build because there's a merge conflict but I
> don't think it had problems with it.
>
> Returning a list-initialized object (and this particular call) is a form
> of copy-list-initialization, which as far as I can tell is directly
> standard-compliant according to the C++ reference
> <https://en.cppreference.com/w/cpp/language/list_initialization>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pioneerspacesim/pioneer/pull/4947#discussion_r482232032>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEQ4NNM3JPGNWOOGLB6T53SDZ4C5ANCNFSM4QN3LVNA>
> .
>

fluffyfreak reviewed (approved):
This is all good from my perspective 👍
